core/engine/src/builtins/intl/date_time_format/mod.rs

```diff
@@ -0,0 +1,825 @@
+    // 31. Else,
+    } else {
+        // a. Let needDefaults be true.
+        // b. If required is date or any, then
+        // i. For each property name prop of « "weekday", "year", "month", "day" », do
+        // 1. Let value be formatOptions.[[<prop>]].
+        // 2. If value is not undefined, set needDefaults to false.
+        // c. If required is time or any, then
+        // i. For each property name prop of « "dayPeriod", "hour", "minute", "second", "fractionalSecondDigits" », do
+        // 1. Let value be formatOptions.[[<prop>]].
+        // 2. If value is not undefined, set needDefaults to false.
+        let needs_defaults = format_options.check_dtf_type(date_time_format_type);
+        // d. If needDefaults is true and defaults is either date or all, then
+        if needs_defaults && defaults != FormatDefaults::Time {
+            // i. For each property name prop of « "year", "month", "day" », do
+            // 1. Set formatOptions.[[<prop>]] to "numeric".
+            format_options.set_date_defaults();
+        }
+        // e. If needDefaults is true and defaults is either time or all, then
+        if needs_defaults && defaults != FormatDefaults::Date {
+            // i. For each property name prop of « "hour", "minute", "second" », do
+            // 1. Set formatOptions.[[<prop>]] to "numeric".
+            format_options.set_time_defaults();
+        }
+        // TODO (nekevss): Do we have access to the localized formats via `icu_datetime`. Is there
+        // a specific API by which this is accessed.
+        //
+        // f. Let formats be resolvedLocaleData.[[formats]].[[<resolvedCalendar>]].
+        // TODO: Support formatMatcher for formatOptions matcher
+        // g. If formatMatcher is "basic", then
+        // i. Let bestFormat be BasicFormatMatcher(formatOptions, formats).
+        // h. Else,
+        // i. Let bestFormat be BestFitFormatMatcher(formatOptions, formats).
+        best_fit_date_time_format(&format_options)?
+    };
+    // 32. Set dateTimeFormat.[[DateTimeFormat]] to bestFormat.
+    // 33. If bestFormat has a field [[hour]], then
+    // a. Set dateTimeFormat.[[HourCycle]] to hc.
+    // 34. Return dateTimeFormat.
+    Ok(JsObject::from_proto_and_data(
+        prototype,
+        DateTimeFormat {
+            locale: resolved_locale,
+            _calendar_algorithm: intl_options.service_options.calendar_algorithm,
+            time_zone,
+            fieldset,
+            bound_format: None,
+        },
+    ))
+}
+
+fn date_time_style_format(
+    date_style: Option<DateStyle>,
+    time_style: Option<TimeStyle>,
+) -> JsResult<CompositeFieldSet> {
+    let mut builder = FieldSetBuilder::default();
+    builder.length = match date_style {
+        Some(DateStyle::Full | DateStyle::Long) => Some(Length::Long),
+        Some(DateStyle::Medium) => Some(Length::Medium),
+        Some(DateStyle::Short) => Some(Length::Short),
+        None => None, // TODO (nekevss): What to do here?
```

Comment:
Read the length from TimeStyle as fallback. If both DateStyle and TimeStyle are undefined, then handle that error case which is an error case anyway